Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs

```diff
@@ -218,53 +217,78 @@ static FieldDescriptor GetField(MessageDescriptor descriptor, string name)
                 return field;
             }
 
-            static Regex ParsePattern(string pattern)
+            static (Regex, Func<Match, string>) ParsePattern(string pattern)
             {
                 var builder = new StringBuilder("^"); // We want to match the whole string
+                Action<Match, StringBuilder> formatBuilder = null;
                 int currentIndex = 0;
+                int groupIndex = 1;
                 while (currentIndex < pattern.Length)
                 {
                     int starStart = pattern.IndexOf('*', currentIndex);
+                    // Handle the literal between the previous star (if any) and the start of the next one.
+                    // If starStart==-1, this is any trailing literal at the end of the pattern.
+                    string literal = pattern.Substring(currentIndex, (starStart == -1 ? pattern.Length : starStart) - currentIndex);
+                    if (literal != "")
+                    {
+                        builder.Append(Regex.Escape(literal));
+                        formatBuilder += (match, sb) => sb.Append(literal);
+                    }
                     if (starStart < 0)
                     {
                         break;
                     }
-                    builder.Append(Regex.Escape(pattern.Substring(currentIndex, starStart - currentIndex)));
+
                     int starEnd = starStart + 1;
                     while (starEnd < pattern.Length && pattern[starEnd] == '*')
                     {
                         starEnd++;
                     }
                     int starCount = starEnd - starStart;
+                    // Deliberately local to have a different variable per 
+                    int matchGroupIndex = groupIndex;
+                    groupIndex++;
+                    // We match any characters within the regex, regardless of number of stars. It just affects
+                    // how the value is escaped. This uses a reluctant matcher, which means x/*/** (which is more likely
+                    // than x/**/*) will match x/a/b/c as "*=a" and then "**=b/c" which is probably the best result.
+                    builder.Append("(.+?)");
                     switch (starCount)
                     {
                         case 1:
-                            builder.Append("[^/]+");
+                            formatBuilder += (match, sb) => sb.Append(Uri.EscapeDataString(match.Groups[matchGroupIndex].Value));
                             break;
                         case 2:
-                            builder.Append(".+");
+                            formatBuilder += (match, sb) => sb.Append(string.Join("/", match.Groups[matchGroupIndex].Value.Split('/').Select(segment => Uri.EscapeDataString(segment))));
                             break;
                         default:
                             throw new ArgumentException($"Resource pattern '{pattern}' is invalid");
```

Comment:
Ah, OK, I see this fails if there are more than 2 starts, but, would it be better if we failed sooner?

Reply:
Could do... this way is simpler if we need significantly different code based on one or two stars. It's not clear to me yet whether we do.

I'll merge the code as-is now, and we can easily refactor later.